### DISCUSSION TITLE
Framework overhead due to ideas

### DISCUSSION THREAD

**Discussion by ghost** [laravel/framework#49987] [Ideas]
I noticed lots of people request features that they use in their projects, to be embedded into laravel. Many times devs are using php functions instead of using the embedded code because they don't know about it and sometimes they don't want to.

**This leads to unused overhead,** that coupled with the **lumen funeral**, leads to bigger bootstrap time (pls don't mention octane because that is not a solution for day to day projects, that is just a marketing stuff to say: we have it, it is fast, but if you have static and singletons... well bad luck).

Can these new features be made in components? So if I don't want for example a component, I should not get it by default?

(especially features that are just a wrapper over php functions: number_format and so on).

**Comment by shaedrich** [comment]
* With the [11.x Slim skeleton](https://github.com/laravel/framework/pull/47309), a lot of overhead has been removed from the framework
* Taylor already is very critical already about what to add to the framework and whether it should better be implemented as a separate package as he states in his official PR rejection template.

**Reply by henzeb** [reply]
Indeed, he showed it at @LaraconEU and it's pretty neat!

Package creators should use deferred providers whenever possible ...

**Reply by Sophist-UK** [reply]
Yes, @taylorotwell  is already careful about what gets added. But equally, as @marius-mcp suggests, there may be big chunks of functionality (e.g. like Queues) which are often not used and which could be split out into a separate and optional package.

There are two reasons for this:

a. The bootstrap overhead - which can be mitigated to some extent by configuration;
b. The attack area - the best form of defence against any sort of attack on unused code is simply not to have the code installed.

The good news is that Laravel is already very modular, so technically this might not be that difficult. However, you wouldn't want to create problems for existing sites upgrading and you might want to make it even easier to install these options than the normal compose functionality i.e. in the Laravel installation wizard and with `artisan` commands, and maybe even make `artisan` include such packages automatically the first time you e.g. create a Queue object with an artisan command.

**Reply by Sophist-UK** [reply]
I haven't yet looked at the "deferred providers" in detail. As far as I can tell, they get autoloaded only when called, which is pretty neat.

The [Laravel 10 documentation](https://laravel.com/docs/10.x/providers#introduction) actually lists several Laravel services that take advantage of "deferred providers" e.g. "mailer, queue, cache" and so perhaps these would also benefit from being split into separate packages. (Separate packages and deferred providers are not IMO mutually exclusive. A package eliminates the code if it is never needed, deferred providers eliminates coding for only those transactions where it is not needed.)

**Reply by shaedrich** [reply]
Well, there's also always the possibility to fork Lumen and maintain it.

But yeah, what I liked about Symfony, is that not everything will be installed with the core package right aways but can be installed when needed.

**Reply by henzeb** [reply]
Seems like the quote reply function broke.

But on deferred providers, due to the fact that they only register when some part of you application needs it, it significantly speeds up the bootstrap and also preserves memory. https://medium.com/@miladev95/laravel-deferred-service-providers-improve-performance-by-deferring-d6a7b174fa4d

Also a thing you can do is utilizing the afterResolving function when registering mixins on facades. Because I think they just fire up the resolving of the object behind the facade class. Which basically undoes the whole deferred thing.

**Reply by henzeb** [reply]
Just wipped up a little macro, which can defer mixins on facade as well as other macroables that you plan to inject using the DI. Would be neat if something like this would be part of the Facade and Macroable classes. Right now I tied it to the App facade. The container is already loaded anyway. It might need some extra checks, for example if the macroable classname actually is macroable.

If this was part of the framework, packages could easily adopt this approach and not becoming the bottleneck of the application.

```php
App::macro(
    'deferMixin',
    function (string $macroable, string $mixin) {

        if (is_a($macroable, Facade::class, true)) {

            $accessor = (fn() => $macroable::getFacadeAccessor())
                ->bindTo(null, Facade::class)();

            $macroable = class_exists($accessor) ? $accessor : $macroable;
        }

        $accessor ??= $macroable;

        $callable = fn() => $macroable::mixin(resolve($mixin));

        if ($this->resolved($accessor)) {
            $callable();
            return;
        }

        $this->afterResolving($accessor, $callable);
    }
);
```

usage: 
````php
App::deferMixin(Illuminate\Support\Facades\Cache::class, CacheMixin::class);
````
